File: servers/fs/initrd/hello.js
// A JavaScript file loaded from the oxbow filesystem.  Run with:  js /hello.js
print("hello from a .js file on oxbow!");

// generators
function* primes(limit) {
  const sieve = new Array(limit).fill(true);
  for (let n = 2; n < limit; n++) {
    if (sieve[n]) {
      yield n;
      for (let m = n * n; m < limit; m += n) sieve[m] = false;
    }
  }
}
print("primes < 40:", [...primes(40)].join(" "));

// destructuring, Map, Set, spread
const [a, b, ...rest] = [1, 2, 3, 4, 5];
print("destructure:", a, b, "rest:", JSON.stringify(rest));
const m = new Map([["x", 1], ["y", 2]]);
print("map x+y:", m.get("x") + m.get("y"));
print("unique:", [...new Set([1, 1, 2, 3, 3, 3])].join(","));

// higher-order + closures
const compose = (f, g) => x => f(g(x));
const inc = x => x + 1, dbl = x => x * 2;
print("compose:", compose(inc, dbl)(10));

// optional chaining + nullish coalescing
const obj = { a: { b: 42 } };
print("opt chain:", obj?.a?.b, obj?.z?.w ?? "default");

// tagged template + reduce
const nums = [1, 2, 3, 4, 5];
print("sum/prod:", nums.reduce((a, b) => a + b), nums.reduce((a, b) => a * b));

File: servers/fs/initrd/hello.py
# A Python script loaded from the oxbow filesystem.  Run with:  py /hello.py
print("hello from a .py file on oxbow!")

# generators
def primes(limit):
    sieve = [True] * limit
    for n in range(2, limit):
        if sieve[n]:
            yield n
            for m in range(n * n, limit, n):
                sieve[m] = False

print("primes < 40:", list(primes(40)))

# slicing, enumerate, dict/set comprehensions
words = "the quick brown fox jumps".split()
print("reversed:", words[::-1])
print("lengths:", {w: len(w) for w in words})
print("unique letters:", len({c for w in words for c in w}))

# a class with operator overloading
class Vec:
    def __init__(self, x, y):
        self.x, self.y = x, y
    def __add__(self, o):
        return Vec(self.x + o.x, self.y + o.y)
    def __repr__(self):
        return "Vec({}, {})".format(self.x, self.y)

print("vector sum:", Vec(1, 2) + Vec(3, 4))

# closures + map/filter
adder = lambda n: lambda x: x + n
add10 = adder(10)
print("map+filter:", list(filter(lambda x: x % 2 == 0, map(add10, range(5)))))
